core/src/main/java/hudson/model/Computer.java

```diff
@@ -939,7 +946,12 @@ public String getSearchUrl() {
      * If this is the master, it returns the system property of the master computer.
      */
     public EnvVars getEnvironment() throws IOException, InterruptedException {
-        return EnvVars.getRemote(getChannel());
+        if (cachedEnvironment != null) {
+            return cachedEnvironment;
```

Comment:
Nope. Just store the value in a local field and return that if non-null.

The cached value should be much like String's hashCode field, in that, as long as non-null it should be equal to a re-run.